File: android/app/src/main/java/com/zipcrack/app/core/ZipCrypto.kt
package com.zipcrack.app.core

import java.io.ByteArrayOutputStream
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.util.zip.CRC32
import java.util.zip.Inflater

class ZipCrackException(message: String) : Exception(message)

/** One encrypted ZipCrypto entry (ciphertext kept for CRC verification). */
class ZipTarget(
    val method: Int,
    val crc32: Long,
    val uncomp: Long,
    val encHeader: ByteArray, // 12 bytes
    val checkByte: Int,
    val data: ByteArray,
)

class ZipArchive(val targets: List<ZipTarget>) {
    fun testPassword(password: String): Boolean {
        if (targets.isEmpty()) return false
        if (!verifyTarget(password, targets[0])) return false
        for (i in 1 until targets.size) {
            if (!verifyTarget(password, targets[i])) return false
        }
        return true
    }

    companion object {
        private const val SIG_LOCAL = 0x04034b50
        private const val SIG_CENTRAL = 0x02014b50
        private const val SIG_EOCD = 0x06054b50

        fun open(raw: ByteArray): ZipArchive {
            val eocd = findEocd(raw)
            val cdOff = u32(raw, eocd + 16)
            val cdSize = u32(raw, eocd + 12)
            val nEntries = u16(raw, eocd + 10)
            if (cdOff.toLong() + cdSize > raw.size) {
                throw ZipCrackException("не ZIP: central directory за пределами файла")
            }

            val targets = ArrayList<ZipTarget>()
            var pos = cdOff
            for (i in 0 until nEntries) {
                if (pos + 46 > raw.size) break
                if (u32(raw, pos) != SIG_CENTRAL) {
                    throw ZipCrackException("не ZIP: битый central directory")
                }
                val flags = u16(raw, pos + 8)
                val method = u16(raw, pos + 10)
                val modTime = u16(raw, pos + 12)
                val crc = u32(raw, pos + 16).toLong() and 0xffffffffL
                val compSize = u32(raw, pos + 20)
                val uncomp = u32(raw, pos + 24).toLong() and 0xffffffffL
                val nameLen = u16(raw, pos + 28)
                val extraLen = u16(raw, pos + 30)
                val commentLen = u16(raw, pos + 32)
                val localOff = u32(raw, pos + 42)
                pos += 46 + nameLen + extraLen + commentLen

                val encrypted = flags and 0x1 != 0
                if (!encrypted) continue
                // AES entries are handled via 7zz — skip here.
                if (method == 99) continue
                if (method != 0 && method != 8) {
                    throw ZipCrackException("метод сжатия $method не поддерживается в native-режиме")
                }
                if (compSize < 12) {
                    throw ZipCrackException("зашифрованная запись слишком короткая")
                }
                targets.add(
                    loadLocal(raw, localOff, method, crc, uncomp, modTime, flags, compSize)
                )
            }

            if (targets.isEmpty()) {
                if (nEntries == 0) throw ZipCrackException("ZIP пуст")
                throw ZipCrackException("нет ZipCrypto-записей для native-проверки")
            }
            targets.sortBy { it.data.size }
            return ZipArchive(targets)
        }

        private fun loadLocal(
            raw: ByteArray,
            off: Int,
            method: Int,
            crc: Long,
            uncomp: Long,
            modTime: Int,
            flags: Int,
            compSize: Int,
        ): ZipTarget {
            if (off < 0 || off + 30 > raw.size) {
                throw ZipCrackException("local header за пределами файла")
            }
            if (u32(raw, off) != SIG_LOCAL) {
                throw ZipCrackException("битый local header")
            }
            val nameLen = u16(raw, off + 26)
            val extraLen = u16(raw, off + 28)
            val dataStart = off + 30 + nameLen + extraLen
            val dataEnd = dataStart + compSize
            if (dataStart < 0 || dataEnd > raw.size || dataStart + 12 > raw.size) {
                throw ZipCrackException("payload за пределами файла")
            }
            val header = raw.copyOfRange(dataStart, dataStart + 12)
            val data = raw.copyOfRange(dataStart + 12, dataEnd)
            val checkByte = if (flags and 0x8 != 0) {
                (modTime shr 8) and 0xff
            } else {
                ((crc shr 24) and 0xff).toInt()
            }
            return ZipTarget(method, crc, uncomp, header, checkByte, data)
        }

        private fun findEocd(raw: ByteArray): Int {
            if (raw.size < 22) throw ZipCrackException("не ZIP-архив")
            val start = raw.size - 22
            val limit = maxOf(0, raw.size - (22 + 65535))
            for (i in start downTo limit) {
                if (u32(raw, i) == SIG_EOCD) return i
            }
            throw ZipCrackException("не ZIP-архив")
        }

        private fun u16(b: ByteArray, off: Int): Int =
            ByteBuffer.wrap(b, off, 2).order(ByteOrder.LITTLE_ENDIAN).short.toInt() and 0xffff

        private fun u32(b: ByteArray, off: Int): Int =
            ByteBuffer.wrap(b, off, 4).order(ByteOrder.LITTLE_ENDIAN).int
    }
}

private class ZipKeys {
    var k0: Int = 0x12345678
    var k1: Int = 0x23456789
    var k2: Int = 0x34567890

    fun init(password: String) {
        k0 = 0x12345678
        k1 = 0x23456789
        k2 = 0x34567890
        for (ch in password) {
            update((ch.code and 0xff).toByte())
        }
    }

    fun update(c: Byte) {
        k0 = crc32Update(k0, c)
        k1 = (k1 + (k0 and 0xff)) * 134775813 + 1
        k2 = crc32Update(k2, ((k1 ushr 24) and 0xff).toByte())
    }

    fun decryptByte(): Int {
        val temp = (k2 or 2).toLong() and 0xffffffffL
        return (((temp * (temp xor 1)) ushr 8) and 0xff).toInt()
    }
}

private val CRC_TABLE: IntArray = IntArray(256).also { t ->
    for (i in 0 until 256) {
        var c = i
        repeat(8) {
            c = if (c and 1 != 0) 0xedb88320.toInt() xor (c ushr 1) else c ushr 1
        }
        t[i] = c
    }
}

private fun crc32Update(crc: Int, b: Byte): Int =
    CRC_TABLE[(crc xor (b.toInt() and 0xff)) and 0xff] xor (crc ushr 8)

private fun verifyTarget(password: String, t: ZipTarget): Boolean {
    val k = ZipKeys()
    k.init(password)
    var last = 0
    for (i in 0 until 12) {
        val c = (t.encHeader[i].toInt() and 0xff) xor k.decryptByte()
        k.update(c.toByte())
        last = c
    }
    if (last != t.checkByte) return false

    val plain = ByteArray(t.data.size)
    for (i in t.data.indices) {
        val c = (t.data[i].toInt() and 0xff) xor k.decryptByte()
        k.update(c.toByte())
        plain[i] = c.toByte()
    }

    val body: ByteArray = when (t.method) {
        0 -> plain
        8 -> {
            try {
                inflateRaw(plain, t.uncomp)
            } catch (_: Exception) {
                return false
            }
        }
        else -> return false
    }

    if (t.uncomp > 0 && body.size.toLong() != t.uncomp) return false
    val crc = CRC32()
    crc.update(body)
    return crc.value == t.crc32
}

private fun inflateRaw(data: ByteArray, uncomp: Long): ByteArray {
    val inflater = Inflater(true)
    inflater.setInput(data)
    val out = ByteArrayOutputStream(if (uncomp > 0) uncomp.toInt().coerceAtLeast(64) else 4096)
    val buf = ByteArray(8192)
    try {
        while (!inflater.finished()) {
            val n = inflater.inflate(buf)
            if (n == 0) {
                if (inflater.needsInput()) break
                if (inflater.needsDictionary()) throw ZipCrackException("dict")
            } else {
                out.write(buf, 0, n)
                if (uncomp > 0 && out.size() > uncomp + 64) break
            }
        }
    } finally {
        inflater.end()
    }
    return out.toByteArray()
}

File: crack/zipcrypto.go
package crack

import (
	"bytes"
	"compress/flate"
	"encoding/binary"
	"errors"
	"fmt"
	"hash/crc32"
	"io"
	"os"
)

// Traditional PKWARE ZipCrypto: decrypt encryption header + payload, then
// verify by inflating (or storing) and matching CRC-32. Header check-byte
// alone is only 8 bits → ~1/256 false positives; CRC closes that gap.

var (
	ErrNotZip       = errors.New("not a ZIP archive")
	ErrNotEncrypted = errors.New("archive is not password-protected")
	ErrUnsupported  = errors.New("unsupported ZIP encryption (need ZipCrypto; AES not supported yet)")
	ErrNoEntries    = errors.New("ZIP has no file entries")
)

const (
	sigLocalFile  = 0x04034b50
	sigCentralDir = 0x02014b50
	sigEndCentral = 0x06054b50
)

// zipTarget is one encrypted entry used for password checks.
type zipTarget struct {
	method    uint16
	crc32     uint32
	uncomp    uint32
	encHeader [12]byte
	checkByte byte
	// Ciphertext after the 12-byte encryption header.
	data []byte
}

// ZipArchive holds pre-parsed encryption targets from a ZIP file.
type ZipArchive struct {
	path    string
	targets []zipTarget
}

// OpenZip parses the archive (via EOCD + central directory) and loads
// encrypted entry ciphertext for verification.
func OpenZip(path string) (*ZipArchive, error) {
	raw, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	return OpenZipBytes(raw, path)
}

// OpenZipBytes is like OpenZip but with already-read bytes.
func OpenZipBytes(raw []byte, path string) (*ZipArchive, error) {
	targets, err := parseZipTargets(raw)
	if err != nil {
		return nil, err
	}
	return &ZipArchive{path: path, targets: targets}, nil
}

func parseZipTargets(raw []byte) ([]zipTarget, error) {
	eocd, err := findEOCD(raw)
	if err != nil {
		return nil, err
	}
	cdOff := binary.LittleEndian.Uint32(raw[eocd+16 : eocd+20])
	cdSize := binary.LittleEndian.Uint32(raw[eocd+12 : eocd+16])
	nEntries := binary.LittleEndian.Uint16(raw[eocd+10 : eocd+12])
	if uint64(cdOff)+uint64(cdSize) > uint64(len(raw)) {
		return nil, fmt.Errorf("%w: central directory out of range", ErrNotZip)
	}

	var targets []zipTarget
	pos := int(cdOff)
	cdEnd := int(cdOff) + int(cdSize)
	for i := 0; i < int(nEntries); i++ {
		if pos+46 > cdEnd && pos+46 > len(raw) {
			return nil, fmt.Errorf("%w: truncated central directory", ErrNotZip)
		}
		if pos+4 > len(raw) {
			break
		}
		sig := binary.LittleEndian.Uint32(raw[pos : pos+4])
		if sig != sigCentralDir {
			return nil, fmt.Errorf("%w: bad central dir signature", ErrNotZip)
		}
		if pos+46 > len(raw) {
			return nil, fmt.Errorf("%w: truncated central header", ErrNotZip)
		}
		flags := binary.LittleEndian.Uint16(raw[pos+8 : pos+10])
		method := binary.LittleEndian.Uint16(raw[pos+10 : pos+12])
		modTime := binary.LittleEndian.Uint16(raw[pos+12 : pos+14])
		crc := binary.LittleEndian.Uint32(raw[pos+16 : pos+20])
		compSize := binary.LittleEndian.Uint32(raw[pos+20 : pos+24])
		uncomp := binary.LittleEndian.Uint32(raw[pos+24 : pos+28])
		nameLen := int(binary.LittleEndian.Uint16(raw[pos+28 : pos+30]))
		extraLen := int(binary.LittleEndian.Uint16(raw[pos+30 : pos+32]))
		commentLen := int(binary.LittleEndian.Uint16(raw[pos+32 : pos+34]))
		localOff := binary.LittleEndian.Uint32(raw[pos+42 : pos+46])
		pos += 46 + nameLen + extraLen + commentLen

		encrypted := flags&0x1 != 0
		if !encrypted {
			continue
		}
		if method == 99 {
			return nil, ErrUnsupported
		}
		if method != 0 && method != 8 {
			return nil, fmt.Errorf("%w: compression method %d", ErrUnsupported, method)
		}
		if compSize < 12 {
			return nil, fmt.Errorf("encrypted entry too small")
		}

		t, err := loadLocalTarget(raw, int(localOff), method, crc, uncomp, modTime, flags, int(compSize))
		if err != nil {
			return nil, err
		}
		targets = append(targets, t)
	}

	if len(targets) == 0 {
		if nEntries == 0 {
			return nil, ErrNoEntries
		}
		return nil, ErrNotEncrypted
	}
	// Prefer smallest payload first (fast reject).
	best := 0
	for i := 1; i < len(targets); i++ {
		if len(targets[i].data) < len(targets[best].data) {
			best = i
		}
	}
	if best != 0 {
		targets[0], targets[best] = targets[best], targets[0]
	}
	return targets, nil
}

func loadLocalTarget(raw []byte, off int, method uint16, crc, uncomp uint32, modTime, flags uint16, compSize int) (zipTarget, error) {
	var t zipTarget
	if off < 0 || off+30 > len(raw) {
		return t, fmt.Errorf("%w: local header out of range", ErrNotZip)
	}
	if binary.LittleEndian.Uint32(raw[off:off+4]) != sigLocalFile {
		return t, fmt.Errorf("%w: bad local header signature", ErrNotZip)
	}
	nameLen := int(binary.LittleEndian.Uint16(raw[off+26 : off+28]))
	extraLen := int(binary.LittleEndian.Uint16(raw[off+28 : off+30]))
	dataStart := off + 30 + nameLen + extraLen
	dataEnd := dataStart + compSize
	if dataStart < 0 || dataEnd > len(raw) || dataStart+12 > len(raw) {
		return t, fmt.Errorf("%w: local payload out of range", ErrNotZip)
	}
	copy(t.encHeader[:], raw[dataStart:dataStart+12])
	t.data = raw[dataStart+12 : dataEnd]
	t.method = method
	t.crc32 = crc
	t.uncomp = uncomp
	if flags&0x8 != 0 {
		t.checkByte = byte(modTime >> 8)
	} else {
		t.checkByte = byte(crc >> 24)
	}
	return t, nil
}

func findEOCD(raw []byte) (int, error) {
	// EOCD is at least 22 bytes; comment max 65535.
	min := 22
	if len(raw) < min {
		return 0, ErrNotZip
	}
	start := len(raw) - min
	limit := len(raw) - (min + 65535)
	if limit < 0 {
		limit = 0
	}
	for i := start; i >= limit; i-- {
		if binary.LittleEndian.Uint32(raw[i:i+4]) == sigEndCentral {
			return i, nil
		}
	}
	return 0, ErrNotZip
}

// TestPassword returns true if password decrypts and CRC-verifies targets.
// Uses the first (smallest) target for speed; confirms with others if present.
func (z *ZipArchive) TestPassword(password string) bool {
	if z == nil || len(z.targets) == 0 {
		return false
	}
	if !verifyTarget(password, &z.targets[0]) {
		return false
	}
	for i := 1; i < len(z.targets); i++ {
		if !verifyTarget(password, &z.targets[i]) {
			return false
		}
	}
	return true
}

func verifyTarget(password string, t *zipTarget) bool {
	// Fast reject via 12-byte header check byte (~1/256 pass).
	k := initKeys(password)
	var last byte
	for i := 0; i < 12; i++ {
		c := t.encHeader[i] ^ k.decryptByte()
		k.update(c)
		last = c
	}
	if last != t.checkByte {
		return false
	}

	// Decrypt payload with continued keys.
	plain := make([]byte, len(t.data))
	for i := 0; i < len(t.data); i++ {
		c := t.data[i] ^ k.decryptByte()
		k.update(c)
		plain[i] = c
	}

	var body []byte
	switch t.method {
	case 0: // store
		body = plain
	case 8: // deflate
		fr := flate.NewReader(bytes.NewReader(plain))
		var buf bytes.Buffer
		// Limit output to declared size + small slack.
		limit := int64(t.uncomp) + 64
		if limit < 64 {
			limit = 1 << 20
		}
		_, err := io.Copy(&buf, io.LimitReader(fr, limit))
		_ = fr.Close()
		if err != nil {
			return false
		}
		body = buf.Bytes()
	default:
		return false
	}

	if t.uncomp > 0 && uint32(len(body)) != t.uncomp {
		return false
	}
	return crc32.ChecksumIEEE(body) == t.crc32
}

// --- ZipCrypto stream cipher ---

type zipKeys [3]uint32

func initKeys(password string) zipKeys {
	k := zipKeys{0x12345678, 0x23456789, 0x34567890}
	for i := 0; i < len(password); i++ {
		k.update(password[i])
	}
	return k
}

func (k *zipKeys) update(c byte) {
	k[0] = crc32Update(k[0], c)
	k[1] = (k[1] + (k[0] & 0xff)) * 134775813 + 1
	k[2] = crc32Update(k[2], byte(k[1]>>24))
}

func (k *zipKeys) decryptByte() byte {
	temp := k[2] | 2
	return byte((temp * (temp ^ 1)) >> 8)
}

// Kept for unit tests of the keystream.
func checkZipCrypto(password string, header [12]byte, checkByte byte) bool {
	k := initKeys(password)
	var last byte
	for i := 0; i < 12; i++ {
		c := header[i] ^ k.decryptByte()
		k.update(c)
		last = c
	}
	return last == checkByte
}

var crcTable = makeCRCTable()

func makeCRCTable() [256]uint32 {
	var t [256]uint32
	for i := 0; i < 256; i++ {
		c := uint32(i)
		for j := 0; j < 8; j++ {
			if c&1 != 0 {
				c = 0xedb88320 ^ (c >> 1)
			} else {
				c >>= 1
			}
		}
		t[i] = c
	}
	return t
}

func crc32Update(crc uint32, b byte) uint32 {
	return crcTable[byte(crc)^b] ^ (crc >> 8)
}
